### ISSUE TITLE
Severe divergence between eager and XLA for tf.math.cosh with reduction ops

### ISSUE THREAD

**Issue by beanduan22** [tensorflow/tensorflow#115508]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.11

### Custom code

Yes

### Python version

3.10

### Current behavior?

TensorFlow produces significantly different results between eager execution and XLA-compiled execution for models involving tf.math.cosh under finite inputs.

In our test, the same model yields differences exceeding 50 in output values between eager and XLA, despite both executions operating on finite inputs and producing finite outputs.

This discrepancy appears to originate from differences in evaluation order or fusion strategies in XLA, and is amplified by reduction operations such as tf.reduce_prod.

Expected behavior:
Eager and XLA execution should produce numerically consistent results within reasonable floating-point tolerance. Differences of this magnitude indicate a potential numerical inconsistency issue that may affect model correctness.

### Standalone code to reproduce the issue

```shell
import os; os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
import tensorflow as tf
import numpy as np

tf.random.set_seed(42)

class CoshBNModel(tf.keras.Model):
    def __init__(self):
        super().__init__()
        self.dense1 = tf.keras.layers.Dense(8, activation='relu')
        self.bn = tf.keras.layers.BatchNormalization()
        self.dense2 = tf.keras.layers.Dense(8, activation='tanh')
        self.ln = tf.keras.layers.LayerNormalization()

    def call(self, x, training=False):
        x = self.dense1(x)
        x = self.bn(x, training=training)
        x = self.dense2(x)
        x = self.ln(x)
        return tf.reduce_prod(tf.math.cosh(x))

model = CoshBNModel()
x_init = tf.random.normal([4, 8])
_ = model(x_init, training=True)

x_base = tf.constant(np.random.RandomState(2).randn(4, 8).astype(np.float32))

def eager_forward(x):
    return model(x, training=False)

@tf.function(jit_compile=True)
def xla_forward(x):
    return model(x, training=False)

# Test at baseline scale
eager_base = eager_forward(x_base)
xla_base = xla_forward(x_base)
diff_base = float(tf.abs(eager_base - xla_base))
print(f"Baseline scale: eager={eager_base.numpy():.6f}, xla={xla_base.numpy():.6f}, diff={diff_base:.3e}")

# Test at large scale (scale_large mutation)
x_large = x_base * 10.0
eager_large = eager_forward(x_large)
xla_large = xla_forward(x_large)
diff_large = float(tf.abs(eager_large - xla_large))
print(f"Large scale: eager={eager_large.numpy():.6f}, xla={xla_large.numpy():.6f}, diff={diff_large:.3e}")

tol = 1e-6
if diff_base > tol or diff_large > tol:
    print("PASS")
else:
    print("NOT REPRODUCED")
```

### Relevant log output

```shell
Baseline scale: eager=150341.765625, xla=150368.218750, diff=2.645e+01
Large scale: eager=97702.742188, xla=97645.078125, diff=5.766e+01
```

**Comment by Venkat6871** [comment]
Hi @beanduan22,

Apologies for the delay, and thanks for the detailed report and reproducible example.

Since this behavior appears with `jit_compile=True`, this looks more like an XLA-related numerical difference than a confirmed correctness bug at this point. XLA does not guarantee identical results to eager execution for floating-point computations, and differences can become more noticeable when combined with downstream ops such as normalization layers and reductions. See:
https://www.tensorflow.org/api_docs/python/tf/function
https://openxla.org/xla/operation_semantics

To help isolate this, I tried a minimal example that compares `tf.math.cosh` directly on the same input (without Dense, BatchNormalization, LayerNormalization, or `tf.reduce_prod`). Sharing a small [gist](https://colab.sandbox.google.com/gist/Venkat6871/549ce00c325abdf27838b358035d6191/115508_tf_2-21-nightly-v.ipynb) for reference.

In that setup, I did not observe a significant difference, which suggests the larger gap in the original script may be coming from accumulated numerical effects in the full model.

Also, if possible, please try with a more recent TensorFlow version, since this report is based on 2.11.

Thanks!

**Comment by beanduan22** [comment]
Thanks for the explanation — I understand that XLA does not guarantee identical floating-point results.

However, the magnitude of the difference here seems larger than typical numerical variation. The discrepancy (~25–50) arises under finite inputs and appears to be amplified by the combination of `cosh` and `reduce_prod`.

While small differences due to evaluation order are expected, this level of divergence suggests that numerical errors may be accumulating or being amplified in a way that could affect model behavior in practice.

Would it be reasonable to consider whether certain patterns (e.g., non-linear ops followed by reductions) might benefit from more numerically stable handling or at least documentation as a known limitation?

Thanks again for taking a look!

**Comment by Kayyuri** [comment]
Hi @beanduan22,
Thank you so much for filing this issue! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!
